File: src/rnet/packets/ZoneVolumePacket.js
const DataPacket = require("./DataPacket");

class ZoneVolumePacket extends DataPacket {
    requiresHandshake() {
        return true;
    }

    getControllerID() {
        return this.sourceControllerID;
    }

    getZoneID() {
        return this.sourcePath[2];
    }

    getVolume() {
        return this.data.readUInt8(0) * 2;
    }
}

ZoneVolumePacket.fromPacket = function(dataPacket) {
    if (dataPacket instanceof DataPacket) {
        const zoneVolumePacket = new ZoneVolumePacket();
        dataPacket.copyToPacket(zoneVolumePacket);
        return zoneVolumePacket;
    }
    else {
        throw new TypeError("Cannot create ZoneVolumePacket with anything other than a DataPacket");
    }
}

module.exports = ZoneVolumePacket;

File: src/RNetPi.Core/RNet/ZoneVolumePacket.cs
using System;

namespace RNetPi.Core.RNet;

/// <summary>
/// Represents an RNet Zone Volume packet
/// </summary>
public class ZoneVolumePacket : DataPacket
{
    public ZoneVolumePacket()
    {
        MessageType = 0x00;
    }

    public byte GetControllerID()
    {
        return SourceControllerID;
    }

    public byte GetZoneID()
    {
        return SourcePath.Length > 2 ? SourcePath[2] : (byte)0;
    }

    public int GetVolume()
    {
        return Data.Length > 0 ? Data[0] * 2 : 0;
    }

    /// <summary>
    /// Creates a ZoneVolumePacket from a DataPacket
    /// </summary>
    public static ZoneVolumePacket FromPacket(DataPacket dataPacket)
    {
        if (dataPacket.MessageType != 0x00)
        {
            throw new ArgumentException("Cannot create ZoneVolumePacket from packet with MessageType != 0x00");
        }

        var zoneVolumePacket = new ZoneVolumePacket();
        dataPacket.CopyToPacket(zoneVolumePacket);
        return zoneVolumePacket;
    }
}